File: Concepts/Heap/basics.go
package main

import (
	"container/heap"
	"fmt"
)

// Define a MinHeap type that implements heap.Interface
type MinHeap []int

// Len is part of sort.Interface. Returns the number of elements in the heap.
func (h MinHeap) Len() int {
	return len(h)
}

// Less is part of sort.Interface. Returns true if the element at index i is smaller than the element at index j.
func (h MinHeap) Less(i, j int) bool {
	return h[i] < h[j] // Min-heap property: smallest element at root
}

// Swap is part of sort.Interface. Swaps the elements at index i and j.
func (h MinHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

// Push is part of heap.Interface. Adds an element to the heap.
// After adding, the heap may violate the heap property, so the heap must "bubble"
// the new element to its correct position.
// This is done by the heap.Push function from the container/heap package,
// which will reorder the heap.
// We append the new element x (which is type-asserted to an int) to the heap.
// The heap package handles the internal reordering (heapifying) to maintain the heap property.
func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(int)) // Type assertion to convert interface{} to int
}

// Pop is part of heap.Interface. Removes and returns the smallest element.
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]     // Get the last element
	*h = old[0 : n-1] // Reduce the slice size
	return x
}

func (h *MinHeap) NthLargest(n int) int {
	tempHeap := &MinHeap{}
	heap.Init(tempHeap)
	for _, num := range *h {
		heap.Push(tempHeap, num)
		if tempHeap.Len() > n {
			heap.Pop(tempHeap)
		}
	}
	return heap.Pop(tempHeap).(int)
}

func (h *MinHeap) NthSmallest(n int) int {
	maxHeap := &MaxHeap{}
	heap.Init(maxHeap)
	for _, num := range *h {
		heap.Push(maxHeap, num)
		if maxHeap.Len() > n {
			heap.Pop(maxHeap)
		}
	}
	return heap.Pop(maxHeap).(int)
}

func (h *MinHeap) Smallest() int {
	if h.Len() == 0 {
		panic("Heap is empty")
	}
	return (*h)[0]
}

func (h *MinHeap) Largest() int {
	if h.Len() == 0 {
		panic("Heap is empty")
	}
	max := (*h)[0]
	for _, val := range *h {
		if val > max {
			max = val
		}
	}
	return max
}

// MaxHeap structure
type MaxHeap []int

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// Main function demonstrating the usage of heap functions
func main() {
	// Create a MinHeap and initialize it
	h := &MinHeap{5, 10, 3, 8}
	heap.Init(h) // Heapify the slice to maintain the heap property

	fmt.Println("Initial heap:", *h) // Output: [3 8 5 10]

	// Push elements into the heap
	heap.Push(h, 2)
	heap.Push(h, 7)
	fmt.Println("Heap after pushes:", *h) // Output: [2 3 5 10 8 7]

	// Pop the smallest element
	fmt.Println("Popped element:", heap.Pop(h)) // Output: 2
	fmt.Println("Heap after pop:", *h)          // Output: [3 7 5 10 8]

	// Remove an element at a specific index
	heap.Remove(h, 2)                      // Removes the element at index 2
	fmt.Println("Heap after removal:", *h) // Output: [3 7 8 10]

	// Push-Pop example
	fmt.Println("Push-Pop result:", heap.PushPop(h, 4)) // Output: 3
	fmt.Println("Heap after Push-Pop:", *h)             // Output: [4 7 8 10]

	// Replace example
	fmt.Println("Replace result:", heap.Replace(h, 6)) // Output: 4
	fmt.Println("Heap after Replace:", *h)             // Output: [6 7 8 10]
}

// Why Use &?
// - The heap package operates on a pointer to the heap because modifying the heap
// 	(e.g., adding or removing elements) requires updating the underlying
// 	data structure in place.
// - By passing a pointer, the methods can directly modify the original heap
// 	instead of working on a copy.

// In Go, when you create a variable like MinHeap{5, 10, 3, 8},
// it creates a value of type MinHeap. If you want to allow modifications to this heap
// from another part of the program, you must work with a pointer.

// Notice the receiver (h *MinHeap) for Push and Pop.
// - The *MinHeap receiver indicates that these methods modify the original heap.
// - If you don't pass a pointer to heap.Push or heap.Pop,
// these methods can't update the actual heap in memory, only a copy of it.

// What Happens Under the Hood?

// Memory Allocation: &MinHeap{5, 10, 3, 8} allocates a MinHeap in memory and
// 					returns a pointer to it.

// Pointer Passing: When you pass h to heap.Init(h) or similar functions, you are
// passing a pointer. Functions like Push and Pop then use this pointer to directly
// modify the original heap.

// Efficiency: Passing a pointer is more efficient than passing large data
// structures by value, especially when dealing with complex or large heaps.

File: Concepts/Heap/basics.py
import heapq

class HeapDemo:
    def __init__(self):
        self.min_heap = []  # Min-heap
        self.max_heap = []  # Max-heap simulated by negating values

    def add_to_min_heap(self, value):
        """Adds a value to the min-heap."""
        heapq.heappush(self.min_heap, value)
        print(f"Added {value} to min-heap: {self.min_heap}")

    def add_to_max_heap(self, value):
        """Adds a value to the max-heap by pushing its negated value."""
        heapq.heappush(self.max_heap, -value)
        print(f"Added {value} to max-heap: {[-val for val in self.max_heap]}")

    def pop_from_min_heap(self):
        """Removes and returns the smallest element from the min-heap."""
        smallest = heapq.heappop(self.min_heap)
        print(f"Removed {smallest} from min-heap: {self.min_heap}")
        return smallest

    def pop_from_max_heap(self):
        """Removes and returns the largest element from the max-heap."""
        largest = -heapq.heappop(self.max_heap)
        print(f"Removed {largest} from max-heap: {[-val for val in self.max_heap]}")
        return largest

    def replace_in_min_heap(self, value):
        """Replaces the smallest element in the min-heap with a new value."""
        replaced = heapq.heapreplace(self.min_heap, value)
        print(f"Replaced {replaced} with {value} in min-heap: {self.min_heap}")
        return replaced

    def pushpop_min_heap(self, value):
        """Pushes a value onto the min-heap and then pops the smallest element."""
        result = heapq.heappushpop(self.min_heap, value)
        print(f"Pushed {value} and popped {result} from min-heap: {self.min_heap}")
        return result

    def heapify_list(self, data):
        """Transforms a list into a heap in-place."""
        heapq.heapify(data)
        print(f"Heapified list: {data}")
        return data

    def get_n_largest(self, n):
        """Returns the n largest elements from the min-heap."""
        largest = heapq.nlargest(n, self.min_heap)
        print(f"{n} largest elements in min-heap: {largest}")
        return largest

    def get_n_smallest(self, n):
        """Returns the n smallest elements from the min-heap."""
        smallest = heapq.nsmallest(n, self.min_heap)
        print(f"{n} smallest elements in min-heap: {smallest}")
        return smallest

# Example usage
def main():
    heap_demo = HeapDemo()

    # Add elements to the min-heap
    heap_demo.add_to_min_heap(10)
    heap_demo.add_to_min_heap(5)
    heap_demo.add_to_min_heap(15)

    # Add elements to the max-heap
    heap_demo.add_to_max_heap(10)
    heap_demo.add_to_max_heap(5)
    heap_demo.add_to_max_heap(15)

    # Remove elements from the heaps
    heap_demo.pop_from_min_heap()
    heap_demo.pop_from_max_heap()

    # Replace the smallest element in the min-heap
    heap_demo.replace_in_min_heap(12)

    # Push a value onto the heap and then pop the smallest element
    heap_demo.pushpop_min_heap(7)

    # Heapify a list
    data = [20, 1, 15, 5, 10]
    heap_demo.heapify_list(data)

    # Get the 2 largest and smallest elements from the min-heap
    heap_demo.get_n_largest(2)
    heap_demo.get_n_smallest(2)

if __name__ == "__main__":
    main()
